nntrainer/include/properties.h

```diff
@@ -0,0 +1,351 @@
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+
+#ifndef __NNTRAINER_PROPERTIES_H__
+#define __NNTRAINER_PROPERTIES_H__
+#ifdef __cplusplus
+
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+template <typename T> struct prop_traits {
+  static constexpr bool is_property = false; /**< check if this is property */
+
+  typedef bool value_type; /**< actual value type of given property */
+
+  /**
+   * @brief validation logic for type T
+   * @param[in] value Target to have registration
+   */
+  static bool is_valid(value_type value) {
+    throw std::runtime_error("no trait type can be validated");
+  }
+
+  /**
+   * @brief get string key of current property
+   */
+  static std::string getKey() {
+    throw std::runtime_error("no trait type have a key");
+  }
+
+  /**
+   * @brief get default value of current type
+   */
+  static value_type getDefault() {
+    throw std::runtime_error("no trait type does not have default");
+  }
+
+  /**
+   * @brief serialization for type T
+   * @param[in] target Target to be converted to string
+   */
+  static std::string serialize(value_type target) {
+    throw std::runtime_error("Only property type can be serialized");
+  }
+
+  /**
+   * @brief deserialization for type T
+   * @param[in] str string to be converted to actual type T
+   * @throw std::runtime_error if T is not property
+   * @throw std::invalid_argument if @a str is not deserializable.
+   */
+  static value_type deserialize(std::string &str) {
+    throw std::runtime_error("Only property type can be deserialized");
+  }
+};
+
+/*****************************************************************************
+ *  Property registration part.                                              *
+ *  If you want to add a property, only this part need change.               *
+ *  You only have to add property struct and prop_traits.                    *
+ *                                                                           *
+ *  It is recommened to use unique key for each types.                       *
+ *  However, it is allowed to use duplicated key                             *
+ *  When those properties are mutually exclusive                             *
+ *****************************************************************************/
+
+/**< color is here for demonstration purpose. */
+
+enum class Color {
+  black,
+  white,
+  unknown,
+};
+
+/**
+ * @brief trait specialization part
+ */
+template <> struct prop_traits<Color> {
+
+  static constexpr bool is_property = true;
+
+  typedef Color value_type;
+
+  static std::string getKey() { return "color"; }
+
+  static bool is_valid(Color c) { return c != Color::unknown; }
+
+  static Color getDefault() { return Color::unknown; }
+
+  static std::string serialize(Color c) {
+    switch (c) {
```

Comment:
lets append append `key` in return of serialize and correspondingly parse it back in de-serialize